tests/bson/bson-enum-003.phpt

```diff
@@ -0,0 +1,108 @@
+--TEST--
+BSON encoding: Enums implementing Persistable
+--SKIPIF--
+<?php require __DIR__ . "/../utils/basic-skipif.inc"; ?>
+<?php skip_if_php_version('<', '8.1.0'); ?>
+--FILE--
+<?php
+
+require_once __DIR__ . '/../utils/basic.inc';
+
+enum MyEnum implements MongoDB\BSON\Persistable
+{
+    case foo;
+
+    public function bsonSerialize(): array
+    {
+        /* Casting the enum to an array is necessary to ensure its properties
+         * are serialized to a BSON document. The "name" property, common to
+         * both pure and backed enums, will be required for unserialization. */
+        return (array) $this;
+    }
+ 
+    public function bsonUnserialize(array $data)
+    {
+        /* Enums do not maintain state, so this method serves no practical
+         * purpose other than being required by the interface. Since an
+         * implementation is required, we will log a message to assert that BSON
+         * decoding always invokes this method. */
+        printf("%s called with: %s\n", __METHOD__, json_encode($data));
+    }
+}
+
+enum MyBackedEnum: string implements MongoDB\BSON\Persistable
+{
+    case foo = 'bar';
+
+    public function bsonSerialize(): array
+    {
+        return (array) $this;
+    }
+ 
+    public function bsonUnserialize(array $data)
+    {
+        printf("%s called with: %s\n", __METHOD__, json_encode($data));
+    }
```

Comment:
We might want to provide a `PersistableEnum` trait to reduce copy-pasta and make this less error-prone for users:
```php
enum MyBackedEnum: string implements Persistable
{
    use PersistableEnum;

    case foo = 'bar';
}
```

Reply:
I see a few options, which may not be mutually exclusive:

 * Allow an enum to be specified in a `typeMap` without requiring it to implement `Unserializable` (since we know the method is irrelevant). In this case, I suppose we'd still invoke `bsonUnserialize()` if it _is_ available on the enum, just to preserve expected behavior for BSON decoding. Since enforcement of `Unserializable` happens elsewhere, this will require some additional code changes.
 * Create an UnserializableEnum trait
 * Create a PersistableEnum trait

Presumably, the traits would exist in PHPC.

Reply:
Added a PersistableEnum trait, which includes a `bsonSerialize()` method that returns the enum object cast as an array and a `bsonUnserialize()` method that does nothing. Presumably, this can be used just fine with an enum implementing any combination of Serializable, Unserializable, and/or Persistable.

If users implement Persistable, the encoded BSON is going to include `__pclass`; otherwise, they'll need to use a typemap and implement Unserializable (as demonstrated in the tests). We'd want to document both approaches, and I suppose the PersistableEnum docs would be a good place to do that.

I don't think there's any reason to create UnserializableEnum, as it'd just be a subset of PersistableEnum with no `bsonSerialize()` method. And if users actually want a different implementation for `bsonSerialize()` (we're well in edge case territory here), they can avoid the trait and just roll their own implementations.

Lastly, I looked into whether we could restrict use of this trait to enum objects (e.g. throw an exception if a non-enum imported the trait) but found no easy way to do so after digging through `zend_compile_use_trait()` in PHP core. I expect this would be possible with some compiler hacking, but I have no idea how to do that and it'd probably be a waste of @sgolemon's time for relatively little benefit.

Reply:
I agree with only providing `PersistableEnum` - also agree with not restricting the use to traits. No matter how much effort you'll put in it, people will always find a way to shoot themselves in the foot.